pkg/backends/influxdb/flush.go

```diff
@@ -0,0 +1,256 @@
+package influxdb
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"strconv"
+	"strings"
+	"sync/atomic"
+
+	"github.com/atlassian/gostatsd"
+)
+
+// flush represents a send operation.
+type flush struct {
+	buffer           *bytes.Buffer
+	writer           io.WriteCloser
+	metricCount      uint64
+	metricsPerBatch  uint64
+	timestampSeconds int64
+	flushIntervalSec float64
+	disabledSubtypes gostatsd.TimerSubtypes
+	errorCounter     *uint64
+	cb               func(buf *bytes.Buffer, seriesCount uint64)
+	getBuffer        func() (*bytes.Buffer, io.WriteCloser)
+	releaseBuffer    func(buf *bytes.Buffer)
+}
+
+// formatNameTags will format a measurement name and tags in an appropriate
+// form for InfluxDB.  The following rules are applied to the keys:
+// - tags with only a value are normalized as "unnamed:value"
+// - if there is a tag of the form "unnamed:x", it will be combined with the regular unnamed values
+// - duplicate values for the same key are sorted (including "unnamed" tags)
+// - keys with multiple values have their values concatenated with __
+// - the final output is sorted by keys
+//
+// Example:
+// ["foo", "key:bar", "unnamed:baz", "key:thing", "other:something"]
+//
+// Process unnamed, sort values, and concatenate to a single value:
+// unnamed: ["foo", "baz"] -> sort as ["baz", "foo"] -> concatenate as "unnamed=baz__foo"
+// key: ["bar", "thing"] -> sort as ["bar", "thing"] -> concatenate as "key=bar__thing"
+// other: ["something"] -> sort as ["something"] -> concatenate as "other=something"
+//
+// Sort by key and merge:
+// key=bar__thing,other=something,unnamed=baz__foo
+//
+// The name will be prepended, and a space will be appended.
+func formatNameTags(name string, tags gostatsd.Tags) string {
+	newTags := make(map[string][]string, len(tags))
+	for _, tag := range tags {
+		kv := strings.SplitN(tag, ":", 2)
+
+		var key, value string
+		if len(kv) == 1 {
+			key = "unnamed"
+			value = kv[0]
+		} else {
+			key = kv[0]
+			value = kv[1]
+		}
+		newTags[key] = append(newTags[key], value)
+	}
+
+	sb := &strings.Builder{}
+	escapeNameToBuilder(sb, name)
+	keys := make([]string, 0, len(newTags))
+	for key, values := range newTags {
+		keys = append(keys, key)
+		sort.Strings(values)
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		sb.WriteByte(',')
+		escapeTagToBuilder(sb, key)
+		sb.WriteByte('=')
+		for idx, value := range newTags[key] {
+			if idx > 0 {
+				sb.WriteString("__")
+			}
+			escapeTagToBuilder(sb, value)
+		}
+	}
+	sb.WriteByte(' ')
+	return sb.String()
+}
+
+// escape will escape NL (\n), CR (\r), TAB (\t), space, comma, backslash, and equals
+// writing the results to the supplied strings.Builder.
+func escapeTagToBuilder(sb *strings.Builder, s string) {
+	for _, ch := range s {
+		if ch == '\n' {
```

Comment:
Would this be more efficient using a switch instead of if-else chain?

Reply:
I'd expect the difference to be down in the noise, if there is one.  Also this is only called on flush, it's not a hot path.